fix: msal-browser not working in non enumerable context (storage)

Bartheleway opened the pull request:
When window storage is being proxied key might not be enumerable anymore. Safe way is to use key() which returns the n-th key of the storage.

This intends to fix #4379.

I did test modifying directly the code into the compiled source and report that it is working as expected.

ghost commented:
Thank you for your contribution! Please generate change files by running `npm run beachball:change` from the root of the repo, then answer each prompt and push up the generated files.

jasonnutter reviewed (changes requested):
Thanks for filing this PR. Please include the following:

1. A test demonstrating the fix.
2. A comment in the file explaining why this change was made.
3. Change file, created by running npm run beachball:change in the root of the repo.

Thanks!

Bartheleway commented:
Hi, thanks for reviewing this quickly

1. I will try to make a test, not sure how I can setup the storageProxy in test as it is code written by Salesforce developers. Do you also want a test that shows that previous version and new works the same when no proxy ?
2. I don't get what you want, the commit message is describing the change and that is the correct place for explaining changes. Code comment should only clarify functionality not change. Do you agree ?
3. Ok, I missed that point, will do

jasonnutter commented:
> 1. I will try to make a test, not sure how I can setup the storageProxy in test as it is code written by Salesforce developers. Do you also want a test that shows that previous version and new works the same when no proxy ?

A test that shows when `window.localStorage` and `window.sessionStorage`  are not enumerable, that relevant APIs still work.

> 2. I don't get what you want, the commit message is describing the change and that is the correct place for explaining changes. Code comment should only clarify functionality not change. Do you agree ?

I disagree. Comments can be used to detail why a piece of code exists or is written in a certain way, in addition to in the commit message.

> 3. Ok, I missed that point, will do

Thanks!

jasonnutter commented on lib/msal-browser/src/cache/BrowserStorage.ts:
```diff
@@ -39,7 +39,13 @@ export class BrowserStorage implements IWindowStorage<string> {
-        return Object.keys(this.windowStorage);
+        const keys: string[] = [];
+
+        for (let i = 0; i < this.windowStorage.length; ++i) {
```
```suggestion
        // Manually iterate (instead of using Object.keys()) to properly support environments where localStorage/sessionStorage are not enumerable, e.g. Salesforce
        for (let i = 0; i < this.windowStorage.length; ++i) {
```

Bartheleway committed:
fix: Object.keys might return false result
When window storage is being proxied key might not be enumerable anymore. Safe way is to use key() which returns the n-th key of the storage.

Bartheleway committed:
fix: containsKey was returning false result in non enumarable context

ghost commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/AzureAD/microsoft-authentication-library-for-js?pullRequest=4380) <br/>All CLA requirements met.

Bartheleway committed:
Change files

ghost commented:
Reminder: The next release is scheduled for next week and this PR appears to be stale :( 

If changes have been requested please address feedback. 
If this PR is still a work in progress please mark as draft.

ghost commented:
Reminder: The next release is scheduled for next week and this PR appears to be stale :( 

If changes have been requested please address feedback. 
If this PR is still a work in progress please mark as draft.

jasonnutter commented:
Closing, per comment in https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4379#issuecomment-1088042572 to not accept fixes for unsupported environments that are not following standards.
